### PR TITLE
Update UPI vSphere terraform code for v0.12

### PR DESCRIPTION
**Pull request by DanyC97**
This PR updates the UPI vSphere terraform templates to v0.12.

Ref https://github.com/openshift/installer/pull/1936#issuecomment-507863827

@abhinavdahiya @staebler @vrutkovs please help review it.

### CODE REVIEW

File: upi/vsphere/README.md [hunk 1]
```diff
@@ -30,30 +31,29 @@ sshKey: YOUR_SSH_KEY
...
 * ipam_token
-* bootstrap_ignition_url
+* bootstrap_ignition_url => The bootstrap ignition config must be placed in a location that will be accessible by the bootstrap machine. For example, you could store the bootstrap ignition config in a gist.
 * control_plane_ignition
 * compute_ignition
-The bootstrap ignition config must be placed in a location that will be accessible by the bootstrap machine. For example, you could store the bootstrap ignition config in a gist.
 
-4. Run `terraform init`.
 
-5. Ensure that you have you AWS profile set and a region specified. The installation will use create AWS route53 resources for routing to the OpenShift cluster.
```
Comment (vrutkovs) [inline upi/vsphere/README.md]: Route53 is still being used for DNS, isn't it?

Comment (DanyC97) [reply upi/vsphere/README.md]: that is correct @vrutkovs is still used hence why i've moved it up in the pre-req section so folks know upfront what is required from the existing code base.

I can extend if y'all okay with to mention why R53 provider is used and not DNS provider, let me know

**Comment by DanyC97** [comment]
job failed due to 
```
govc: Post https://vcsa-ci.vmware.devcluster.openshift.com/sdk: dial tcp 139.178.73.3:443: i/o timeout
2019/07/22 15:21:01 Container setup in pod e2e-vsphere failed, exit code 1, reason Error
```

sounds like you not able to import the template into your vCenter env

**Comment by abhinavdahiya** [comment]
/hold

we are close (almost in) feature freeze, cannot make these changes until there's an exception.

**Comment by DanyC97** [comment]
> /hold
> 
> we are close (almost in) feature freeze, cannot make these changes until there's an exception.

no problem at all, totally understand ! let me know when we are out the freeze window, happy to carry on from there

File: upi/vsphere/main.tf [hunk 2]
```diff
@@ -1,95 +1,97 @@
...
+  name = var.vsphere_datacenter
 }
 
 module "folder" {
   source = "./folder"
 
-  path          = "${var.cluster_id}"
-  datacenter_id = "${data.vsphere_datacenter.dc.id}"
+  //path          = "${var.cluster_id}"
+  path          = "Dani/${var.vm_folder}"
```
Comment (fherbert) [inline upi/vsphere/main.tf]: Is 'Dani/' prepended to the folder path why the tests fail? You probably don't want this included in the PR?

File: upi/vsphere/route53/main.tf [hunk 3]
```diff
@@ -35,58 +35,58 @@ resource "aws_route53_record" "api-external" {
...
-  zone_id = "${aws_route53_zone.cluster.zone_id}"
+  zone_id = aws_route53_zone.cluster.zone_id
   name    = "*.apps.${var.cluster_domain}"
-  records = ["${var.compute_ips}"]
+  records = [var.compute_ips]
 }
 
 resource "aws_route53_record" "control_plane_nodes" {
-  count = "${var.control_plane_count}"
+  count = var.control_plane_count}
```
Comment (fherbert) [inline upi/vsphere/route53/main.tf]: need to remove this trailing '}'

Comment (DanyC97) [reply upi/vsphere/route53/main.tf]: thanks @fherbert , fixed

File: upi/vsphere/main.tf (hunk 2 above)
Comment (DanyC97) [reply upi/vsphere/main.tf]: thanks @fherbert , fixed

**Comment by DanyC97** [comment]
@abhinavdahiya i saw you started to work on https://github.com/openshift/installer/pull/2396 , can i assume we can re-activate this PR ?

**Comment by DanyC97** [comment]
@abhinavdahiya gentle reminder, please let me know if you are still interested in this PR now that 4.2 is out.

If not then please close this PR.

**Comment by mkelnermishal** [comment]
Hey,
Is there any update on this code ? we need it in order to release 4.2 to our customers (VMware).
Thanks.

**Comment by DanyC97** [comment]
> Hey,
> Is there any update on this code ? we need it in order to release 4.2 to our customers (VMware).
> Thanks.

@mkelnermishal happy to continue the work if @abhinavdahiya confirm he is ready/ happy to accept the PR, haven't received any feedback on my previous attempts to get an answer hence my assumption is that is not needed

**Comment by abhinavdahiya** [comment]
Thanks for the changes! And I think it's important that we move to 0.12 terraform..

So here's the pickle, the baremetal using packet.net [here](https://github.com/openshift/installer/tree/master/upi/metal) also uses the terraform..

and the image we use to test both is 0.11 https://github.com/openshift/installer/blob/9dcd4d2f7f9a086f48924c590f7160e7a550b900/images/installer/Dockerfile.upi.ci#L29

so we will have to either make sure:
A) update the CI image used for terraform to 0.12, make sure baremetal have terraform version set to <=0.11 _ i think that should allow 0.12 terraform binary to run in 0.11 mode_ and then we can merge this change.

B) we update all the changes in one PR.

**Comment by DanyC97** [comment]
> A) update the CI image used for terraform to 0.12, make sure baremetal has terraform version set to <=0.11 _i think that should allow 0.12 terraform binary to run in 0.11 mode_ and then we can merge this change.
> 
> B) we update all the changes in one PR.

@abhinavdahiya thank you for taking the time to respond and refocus attention on this PR. I'm easy going with either of the 2 suggestions.

What does the baremetal team prefer to go with?

**Comment by DanyC97** [comment]
maybe merge the effort in one PR with https://github.com/openshift/installer/pull/2479 or chain them one after another or just be aware of it ;)

**Comment by abhinavdahiya** [comment]
> > A) update the CI image used for terraform to 0.12, make sure baremetal has terraform version set to <=0.11 _i think that should allow 0.12 terraform binary to run in 0.11 mode_ and then we can merge this change.
> > B) we update all the changes in one PR.
> 
> @abhinavdahiya thank you for taking the time to respond and refocus attention on this PR. I'm easy going with either of the 2 suggestions.
> 
> What does the baremetal team prefer to go with?

I would prefer the option B. if you can help migrate the metal work with vsphere work to 0.12. I can help review and get it merged...

**Comment by DanyC97** [comment]
closing in favour of https://github.com/openshift/installer/pull/3235 and https://github.com/openshift/installer/pull/3429

**Pull request closed without merging**